atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -389,26 +429,42 @@ def do_worker_build(self, release, cluster_info, koji_upload_dir, task_id):
                 except OsbsException:
                     pass
 
+    def select_and_start_cluster(self, platform):
+        ''' Choose a cluster and start a build on it '''
+
+        config = get_config(self.workflow)
+        clusters = config.get_enabled_clusters_for_platform(platform)
+
+        if not clusters:
+            raise UnknownPlatformException('No clusters found for platform {}!'
+                                           .format(platform))
+
+        possible_clusters = None
+        retry_info = self.ClusterRetryContexts(clusters, self.max_cluster_fails)
+
+        while True:
+            if not clusters:
+                raise RuntimeError('Could not find appropriate cluster for worker build.')
+            possible_clusters = self.get_clusters(clusters, platform, retry_info)
+            for cluster in possible_clusters:
+                try:
+                    self.do_worker_build(self.release, cluster, self.koji_upload_dir,
+                                         self.fs_task_id)
+                    return
+                except BuildCanceledException:
+                    raise
+                except Exception:
+                    if not retry_info.try_again_later(cluster, self.find_cluster_retry_delay):
+                        clusters.remove(cluster)
+
     def run(self):
-        release = self.get_release()
         platforms = self.get_platforms()
-        koji_upload_dir = self.get_koji_upload_dir()
-        task_id = self.get_fs_task_id()
 
         thread_pool = ThreadPool(len(platforms))
-        result = thread_pool.map_async(
-            lambda cluster_info: self.do_worker_build(release, cluster_info,
-                                                      koji_upload_dir, task_id),
-            [self.choose_cluster(platform) for platform in platforms]
-        )
+        result = thread_pool.map_async(self.select_and_start_cluster, platforms)
 
         try:
-            while not result.ready():
-                # The wait call is a blocking call which prevents signals
-                # from being processed. Wait for short intervals instead
-                # of a single long interval, so build cancellation can
-                # be handled virtually immediately.
-                result.wait(1)
+            result.get()
         # Always clean up worker builds on any error to avoid
         # runaway worker builds (includes orchestrator build cancellation)
         except Exception:
```

Comment:
The handling in this block is no longer correct.
Previously, `Exception` really meant `BuildCanceledException`. 
This is no longer true with the usage of `result.get` since it'll re-raise any exception in the worker threads.

Also, I'm not sure if something happens in a worker thread, the remaining worker builds will carry on as it did previously.

Reply:
I added a terminate() statement to address this